File: src/Block.cpp
#include <Block.hpp>
#include <BlockType.hpp>
#include <Config.hpp>

Block::Block() {
	type = BlockType::random();
	position = sf::Vector2i(GRID_WIDTH / 2 - type->grid[0].size() / 2, 0);
	rotation_state = 0;
}

std::vector<sf::Vector2i> Block::get_tiles() {
	std::vector<sf::Vector2i>tiles = {};
	for (int y = 0; y < type->grid.size(); y++) {
		for (int x = 0; x < type->grid[y].size(); x++) {
			if (!type->grid[y][x]) {
				continue;
			}
			int rotated_x = x;
			int rotated_y = y;
			if (type->rotate) {
				int center_x = type->grid[0].size() / 2;
				int center_y = type->grid.size() / 2;
				int offset_x = x - center_x;
				int offset_y = y - center_y;
				switch (rotation_state) {
					case 0:
						rotated_x = x;
						rotated_y = y;
						break;
					case 1:
						rotated_x = center_x + offset_y;
						rotated_y = center_y - offset_x;
						break;
					case 2:
						rotated_x = center_x - offset_x;
						rotated_y = center_y - offset_y;
						break;
					case 3:
						rotated_x = center_x - offset_y;
						rotated_y = center_y + offset_x;
						break;
					default:
						rotation_state %= 4;
				}
			}
			int global_x = rotated_x + position.x;
			int global_y = rotated_y + position.y;
			tiles.push_back(sf::Vector2i(global_x, global_y));
		}
	}
	return tiles;
}

File: src/structs/block.rs
use crate::config::GRID_WIDTH;
use crate::structs::BlockType;
use sfml::system::Vector2i;

pub struct Block<'a> {
    pub block_type: &'a BlockType,
    pub position: Vector2i,
    pub rotation_state: i32,
}

impl<'a> Block<'a> {
    pub fn new(block_type: &'a BlockType) -> Self {
        Self {
            block_type,
            position: Vector2i::new(
                GRID_WIDTH as i32 / 2 - block_type.grid[0].len() as i32 / 2,
                0,
            ),
            rotation_state: 0,
        }
    }

    pub fn get_tiles(&mut self) -> Vec<Vector2i> {
        let mut tiles: Vec<Vector2i> = Vec::new();
        for (y, row) in self.block_type.grid.iter().enumerate() {
            let y = y as i32;
            for (x, cell) in row.iter().enumerate() {
                let x = x as i32;
                if !cell {
                    continue;
                }
                let mut rotated = Vector2i::new(x as i32, y as i32);
                if self.block_type.rotate {
                    let center_x = row.len() as i32 / 2;
                    let center_y = self.block_type.grid.len() as i32 / 2;
                    let offset_x = x - center_x;
                    let offset_y = y - center_y;
                    match self.rotation_state {
                        0 => {}
                        1 => {
                            rotated.x = center_x + offset_y;
                            rotated.y = center_y - offset_x;
                        }
                        2 => {
                            rotated.x = center_x - offset_x;
                            rotated.y = center_y - offset_y;
                        }
                        3 => {
                            rotated.x = center_x - offset_y;
                            rotated.y = center_y + offset_x;
                        }
                        _ => self.rotation_state %= 4,
                    }
                }
                let global = self.position + rotated;
                tiles.push(global);
            }
        }
        tiles
    }
}
